packages/core/src/new-api/internal/execution/executionStateReducer.ts

```diff
@@ -120,6 +122,18 @@ export function executionStateReducer(
     );
   }
 
+  if (isExecutionFailure(action)) {
```

Comment:
Actually update the execution state on failure. The change to check whether a batch is complete by looking at execution state exposed that this was missing.